client/extensions/woocommerce/lib/generate-variations/index.js

```diff
@@ -0,0 +1,39 @@
+// http://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
+const f = ( a, b ) => [].concat( ...a.map( c => b.map( d => [].concat( c, d ) ) ) );
+const cartesian = ( a, b, ...c ) => b ? cartesian( f( a, b ), ...c ) : a;
+
+/**
+ * Generates variation objects based on a product's attributes.
+ *
+ * @param {Object} product Product object.
+ * @return {Array} Array of variation objects.
+ */
+export default function generateVariations( product ) {
+	const { attributes } = product;
+	const variationTypes = ( attributes && attributes.filter( attribute => attribute.variation ) ) || [];
+
+	if ( variationTypes.length < 2 ) {
```

Comment:
1) I thought we could just short circuit because if there is a single type, the variations are just that type. 2) I thought the API response for variations with one single attribute was slightly different. 

I just went back and checked number 2 and that turned out to be wrong so this doesn't seem as necessary.